File: 7-multiple-deallocation/code.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>

// Struktur data sederhana untuk digunakan
typedef struct {
    int id;
    char* name;
    float score;
} Student;

// Fungsi untuk membuat student baru
Student* create_student(int id, const char* name, float score) {
    Student* s = (Student*)malloc(sizeof(Student));
    if (s == NULL) {
        return NULL;
    }
    
    s->id = id;
    s->name = (char*)malloc(strlen(name) + 1);
    if (s->name == NULL) {
        free(s);
        return NULL;
    }
    
    strcpy(s->name, name);
    s->score = score;
    return s;
}

// Fungsi untuk membebaskan memori student
void free_student(Student* s) {
    if (s != NULL) {
        free(s->name);
        free(s);
    }
}

// Fungsi untuk mencari student berdasarkan ID
void find_student(Student* students[], int count, int target_id) {
    for (int i = 0; i < count; i++) {
        if (students[i]->id == target_id) {
            printf("Found student: %d - %s (Score: %.2f)\n", 
                   students[i]->id, students[i]->name, students[i]->score);
            return;
        }
    }
    printf("Student with ID %d not found.\n", target_id);
}

// Fungsi yang menyebabkan multiple deallocation
void process_student_records(Student* students[], int count) {
    // Memproses data student
    for (int i = 0; i < count; i++) {
        printf("Processing student: %d - %s (Score: %.2f)\n", 
               students[i]->id, students[i]->name, students[i]->score);
        
        // Modifikasi data
        students[i]->score += 10.0;
    }
    
    // Mencari beberapa student
    find_student(students, count, 101);
    find_student(students, count, 105);
    
    // Bug: membebaskan student[0] dua kali
    // Pertama di sini
    free_student(students[0]);
    
    // Kemudian, membebaskan semua records
    for (int i = 0; i < count; i++) {
        // Bug: student[0] akan dibebaskan lagi di sini
        free_student(students[i]);
    }
}

int main() {
    const int NUM_STUDENTS = 5;
    Student* class_roster[5];
    clock_t start, end;
    double cpu_time_used;
    
    // Inisialisasi data
    class_roster[0] = create_student(101, "Alice", 85.5);
    class_roster[1] = create_student(102, "Bob", 78.0);
    class_roster[2] = create_student(103, "Charlie", 92.5);
    class_roster[3] = create_student(104, "Diana", 88.0);
    class_roster[4] = create_student(105, "Edward", 79.5);
    
    start = clock();
    
    // Proses data yang akan menyebabkan multiple deallocation
    process_student_records(class_roster, NUM_STUDENTS);
    
    end = clock();
    cpu_time_used = ((double) (end - start)) / CLOCKS_PER_SEC;
    printf("\nWaktu eksekusi: %f detik\n", cpu_time_used);
    
    return 0;
}

File: 7-multiple-deallocation/code.rs
use std::rc::Rc;
use std::time::Instant;

// Struktur data student
struct Student {
    id: i32,
    name: String,
    score: f32,
}

// Implementasi metode untuk Student
impl Student {
    fn new(id: i32, name: &str, score: f32) -> Self {
        Student {
            id,
            name: String::from(name),
            score,
        }
    }
}

// Fungsi untuk mencari student berdasarkan ID
fn find_student(students: &Vec<Rc<Student>>, target_id: i32) {
    for student in students {
        if student.id == target_id {
            println!("Found student: {} - {} (Score: {:.2})",
                     student.id, student.name, student.score);
            return;
        }
    }
    println!("Student with ID {} not found.", target_id);
}

// Fungsi yang mencoba menyebabkan multiple deallocation
fn process_student_records(students: &mut Vec<Rc<Student>>) {
    // Memproses data student
    for student in students.iter() {
        println!("Processing student: {} - {} (Score: {:.2})", 
                 student.id, student.name, student.score);
        
        // Catatan: Score tidak dapat dimodifikasi karena Student di-wrap dalam Rc
        // dan bersifat immutable
    }
    
    // Mencari beberapa student
    find_student(students, 101);
    find_student(students, 105);
    
    // Mencoba untuk "drop" student[0] secara manual
    // Tidak masalah karena Rc mengelola reference counting
    if !students.is_empty() {
        // Dalam Rust, ini hanya mengurangi reference count, bukan menghapus memori
        let first_student = students[0].clone();
        drop(first_student); // Mengurangi reference count
    }
    
    // Mencoba drop semua students
    // Dalam Rust, ini aman karena sistem kepemilikan
    for i in 0..students.len() {
        let student = students[i].clone();
        drop(student); // Mengurangi reference count
    }
}

fn main() {
    // Membuat vector Rc<Student> untuk berbagi kepemilikan
    let mut class_roster: Vec<Rc<Student>> = vec![
        Rc::new(Student::new(101, "Alice", 85.5)),
        Rc::new(Student::new(102, "Bob", 78.0)),
        Rc::new(Student::new(103, "Charlie", 92.5)),
        Rc::new(Student::new(104, "Diana", 88.0)),
        Rc::new(Student::new(105, "Edward", 79.5)),
    ];
    
    let start = Instant::now();
    
    // Proses data - tidak akan menyebabkan double free karena Rust's ownership system
    process_student_records(&mut class_roster);
    
    let duration = start.elapsed();
    println!("\nWaktu eksekusi: {:?}", duration);
    
    // Semua resource akan dibersihkan secara otomatis saat scope berakhir
    println!("Program completed successfully");
}